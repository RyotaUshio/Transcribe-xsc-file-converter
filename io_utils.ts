import * as fs from "fs"
import * as util from "util"

const readFile = util.promisify(fs.readFile)

// the following function is from https://stackoverflow.com/a/54565854
async function readStreamContents(stream: NodeJS.ReadStream): Promise<string> {
  const chunks = []
  for await (const chunk of stream) chunks.push(chunk)
  return Buffer.concat(chunks).toString("utf8")
}

function readAllOfStdin(): Promise<string> {
  return readStreamContents(process.stdin)
}

async function readAllOfAFile(filePath: string): Promise<string> {
  try {
    const testFileBuffer = await readFile(filePath)
    return testFileBuffer.toString()
  } catch (err) {
    throw err
  }
}

export type InputSource = { type: "stdin" } | { type: "file"; filePath: string }
export async function readAllFromSource(sourceDescriptor: InputSource): Promise<string> {
  if (sourceDescriptor.type === "stdin") {
    return readAllOfStdin()
  } else {
    return readAllOfAFile(sourceDescriptor.filePath)
  }
}
