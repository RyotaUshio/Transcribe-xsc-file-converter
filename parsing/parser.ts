import { splitIntoLines, splitLineIntoParts, getKeyOfLine, lineIsBlank } from "./lex"
import { ParseData } from "./parse_types"
import { initialParseState } from "./initial_parse_state"
import { lineParsers } from "./line_parsers"

function warnIfUnknownVersion(versionLine: string): void {
  if (!versionLine.startsWith("Transcribe!,")) {
    throw new Error("file does not appear to a Transcribe! file")
  }

  const knownVersions = ["Transcribe!,Macintosh OS-X,8,21,7,S,3", "Transcribe!,Macintosh OS-X,8,50,7,S,0"]
  if (!knownVersions.includes(versionLine)) {
    console.error(
      `Warning: file format has version “${versionLine}”, which not a version this program has even been tested against. Data may be wrong or missing.`
    )
  }
}

function parse(transcribeFileContents: string): ParseData {
  let parseState = { ...initialParseState }

  const lines = splitIntoLines(transcribeFileContents)
  const versionLine = lines[1]
  warnIfUnknownVersion(versionLine)

  for (const line of lines.slice(2)) {
    if (lineIsBlank(line)) continue

    const lineKey = getKeyOfLine(line)
    const lineParser = lineParsers[lineKey]
    if (lineParser) {
      parseState = lineParser(line, parseState)
    }
  }

  // TODO include implicit subdivision markers in the final list of markers, probably distinguished by `markerType`.
  // This is most easily done with post-processing, when it’s easy to look at the timestamp of the next marker.

  return parseState.data
}

export { parse }
