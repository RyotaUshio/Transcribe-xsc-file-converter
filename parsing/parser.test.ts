import { parse } from "./parser"
import { readAllOfAFile } from "../io_utils"

test("the file ‘Windows ME Startup.xsc’ can be parsed", async () => {
  const fileContents = await readAllOfAFile("tests/relatively few markers/Windows ME Startup.xsc")
  const parsed = parse(fileContents)
  expect(parsed).toMatchSnapshot()
})
