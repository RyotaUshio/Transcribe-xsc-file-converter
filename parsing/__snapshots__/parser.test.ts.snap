// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the file ‘Windows ME Startup.xsc’ can be parsed 1`] = `
Object {
  "markers": Object {
    "autonumbering": Object {
      "byMarkerType": Object {
        "beat": false,
        "measure": true,
        "section": true,
      },
      "numberContinuously": true,
    },
    "list": Array [
      Object {
        "label": Object {
          "label": "A1",
          "labelIsAutoNamed": true,
        },
        "markerType": Object {
          "name": "section",
          "relativeScope": 3,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 0.167,
          "string": "0:00:00.167",
        },
      },
      Object {
        "label": null,
        "markerType": Object {
          "name": "beat",
          "relativeScope": 1,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 1.039,
          "string": "0:00:01.039",
        },
      },
      Object {
        "label": Object {
          "label": "A2",
          "labelIsAutoNamed": true,
        },
        "markerType": Object {
          "name": "measure",
          "relativeScope": 2,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 1.961,
          "string": "0:00:01.961",
        },
      },
      Object {
        "label": null,
        "markerType": Object {
          "name": "beat",
          "relativeScope": 1,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 2.678,
          "string": "0:00:02.678",
        },
      },
      Object {
        "label": null,
        "markerType": Object {
          "name": "beat",
          "relativeScope": 1,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 3.343,
          "string": "0:00:03.343",
        },
      },
      Object {
        "label": null,
        "markerType": Object {
          "name": "beat",
          "relativeScope": 1,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 4,
          "string": "0:00:04.000",
        },
      },
      Object {
        "label": null,
        "markerType": Object {
          "name": "beat",
          "relativeScope": 1,
        },
        "numSubdivisionsBetweenThisAndNextMarker": Object {
          "type": "same_as_previous_marker",
        },
        "timestamp": Object {
          "hours": 0,
          "minutes": 0,
          "seconds": 4.637,
          "string": "0:00:04.637",
        },
      },
    ],
  },
  "soundFile": Object {
    "audioDurationInSeconds": "5.511837",
    "filePath": "/Users/roryokane/Music/iTunes/iTunes Music/Music/Unknown Artist/Unknown Album/Windows ME Startup.mp3",
    "filename": "Windows ME Startup.mp3",
  },
}
`;
